utils/stall_free.hh

```diff
@@ -56,14 +58,93 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+template <typename T>
+concept GentlyClearable = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.reset() } -> std::same_as<void>;
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept VectorLike = requires (T x, size_t n) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept Container = !VectorLike<T> && requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.erase(x.begin()) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <Container T>
+future<> clear_gently(T& c) noexcept;
+
+template <GentlyClearable T>
+future<> clear_gently(T& o) noexcept(noexcept(o.clear_gently())) {
+    return o.clear_gently();
+}
+
+template <typename T>
+future<> clear_gently(T&) noexcept {
+    return make_ready_future<>();
+}
+
+template <SmartPointer Ptr>
+future<> clear_gently(Ptr& o) {
+    return clear_gently(*o).finally([&o] {
+        return o.reset();
+    });
+}
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept {
+    return do_for_each(a, [] (T& o) {
+        return clear_gently(o);
+    });
+}
+
+// Trivially destructible elements can be safely cleared in bulk
+template <VectorLike T>
+requires std::is_trivially_destructible_v<typename T::value_type>
+future<> clear_gently(T& v) noexcept {
+    v.clear();
```

Comment:
Unless I'm missing something, your VectorLike concept didn't require a clear(), but now you use one.
I think you should have a concept of Clearable - something that has a clear() function, or evenr a TriviallyClearable concept which combines having a clear() and knowing that it's efficient.

Reply:
Good idea. Will do.

Reply:
I didn't do that eventually.
It's not required right now, can add it later.